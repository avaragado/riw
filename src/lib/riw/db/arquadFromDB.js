// @flow

import toPairs from 'ramda/src/toPairs';
import chain from 'ramda/src/chain';
import map from 'ramda/src/map';

const quadify = obj => chain(
    ([key, val]) => {
        if (typeof val === 'string') {
            return [[key, val]];
        }

        return map(
            keys => [key, ...keys],
            quadify(val),
        );
    },
    toPairs(obj),
);

export default (db: RIWDB): RIWDBQuad[] => quadify(db.data);
