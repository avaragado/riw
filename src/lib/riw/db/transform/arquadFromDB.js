// @flow

import toPairs from 'ramda/src/toPairs';
import chain from 'ramda/src/chain';
import map from 'ramda/src/map';

import type { TranslationQuad } from '../../../../types';
import type { TranslationsDB } from '../';

const quadify = obj => chain(
    ([key, val]) => {
        if (typeof val === 'string') {
            return [[key, val]];
        }

        return map(
            keys => [key, ...keys],
            quadify(val),
        );
    },
    toPairs(obj),
);

export default (db: TranslationsDB): TranslationQuad[] => quadify(db.data);
