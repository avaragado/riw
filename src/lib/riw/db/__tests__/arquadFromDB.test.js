// @flow

import arquadFromDB from '../arquadFromDB';

type Fixture = {
    name: string,
    in: RIWDB,
    out: RIWDBQuad[],
};

const fixtures: Fixture[] = [
    {
        name: '01 empty db',
        in: {
            version: 1,
            data: {},
        },
        out: [],
    },

    {
        name: '02 lots of entries',
        in: {
            version: 1,
            data: {
                'hello': {
                    'desc1': {
                        'en-re': 'olleh',
                        'en-UP': 'HELLO',
                    },
                    'desc2': {
                        'fr-fr': 'bonjour',
                    },
                },
                'goodbye': {
                    'desc1': {
                        'en-re': 'eybdoog',
                        'en-UP': 'GOODBYE',
                    },
                },
            },
        },
        out: [
            ['hello', 'desc1', 'en-re', 'olleh'],
            ['hello', 'desc1', 'en-UP', 'HELLO'],
            ['hello', 'desc2', 'fr-fr', 'bonjour'],
            ['goodbye', 'desc1', 'en-re', 'eybdoog'],
            ['goodbye', 'desc1', 'en-UP', 'GOODBYE'],
        ],
    },
];

describe('lib/riw/db/arquadFromDB', () => {
    fixtures.forEach((fixture) => {
        it(fixture.name, () => {
            const received = arquadFromDB(fixture.in);

            expect(received).toEqual(fixture.out);
        });
    });
});
