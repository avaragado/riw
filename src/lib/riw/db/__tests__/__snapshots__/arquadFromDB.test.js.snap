// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/riw/db/arquadFromDB 01 empty db 1`] = `Array []`;

exports[`lib/riw/db/arquadFromDB 02 lots of entries 1`] = `
Array [
  Array [
    "hello",
    "desc1",
    "en-re",
    "olleh",
  ],
  Array [
    "hello",
    "desc1",
    "en-UP",
    "HELLO",
  ],
  Array [
    "hello",
    "desc2",
    "fr-fr",
    "bonjour",
  ],
  Array [
    "goodbye",
    "desc1",
    "en-re",
    "eybdoog",
  ],
  Array [
    "goodbye",
    "desc1",
    "en-UP",
    "GOODBYE",
  ],
]
`;
